twilio/jwt/__init__.py

```diff
@@ -1,83 +1,149 @@
-    return '.'.join(segments)
-
-
-def decode(jwt, key='', verify=True):
-    try:
-        signing_input, crypto_segment = jwt.rsplit('.', 1)
-        header_segment, payload_segment = signing_input.split('.', 1)
-    except ValueError:
-        raise DecodeError("Not enough segments")
-    try:
-        header_raw = base64url_decode(binary(header_segment)).decode('utf-8')
-        payload_raw = base64url_decode(binary(payload_segment)).decode('utf-8')
-        header = json.loads(header_raw)
-        payload = json.loads(payload_raw)
-        signature = base64url_decode(binary(crypto_segment))
-    except (ValueError, TypeError):
-        raise DecodeError("Invalid segment encoding")
-    if verify:
+class Jwt(object):
+    """Base class for building a Json Web Token"""
+    def __init__(self, secret_key, issuer, subject=None, algorithm='HS256', nbf=None,
+                 ttl=3600, valid_until=None):
+        self.secret_key = secret_key
+        """:type str: The secret used to encode the JWT"""
+        self.issuer = issuer
+        """:type str: The issuer of this JWT"""
+        self.subject = subject
+        """:type str: The subject of this JWT, ommited from payload by default"""
+        self.algorithm = algorithm
+        """:type str: The algorithm used to encode the JWT, defaults to 'HS256'"""
+        self.nbf = nbf
+        """:type int: Time in secs since epoch before which this JWT is invalid. Defaults to now."""
+        self.ttl = ttl
+        """:type int: Time to live of the JWT in seconds, defaults to 1 hour"""
+        self.valid_until = valid_until
+        """:type int: Time in secs since epoch this JWT is valid for. Overrides ttl if provided."""
+
+        self.__decoded_payload = None
+        self.__decoded_headers = None
+
+    def _generate_payload(self):
+        """:rtype: dict the payload of the JWT to send"""
+        raise NotImplementedError('Subclass must provide a payload.')
+
+    def _generate_headers(self):
+        """:rtype dict: Additional headers to include in the JWT, defaults to an empty dict"""
+        return {}
+
+    @classmethod
+    def _from_jwt(cls, headers, payload, key=None):
+        """
+        Class specific implementation of from_jwt which should take jwt components and return
+        and instance of this Class with jwt information loaded.
+        :return: Jwt object containing the headers, payload and key
+        """
+        jwt = Jwt(
+            secret_key=key,
+            issuer=payload.get('iss', None),
+            subject=payload.get('sub', None),
+            algorithm=headers.get('alg', None),
+            valid_until=payload.get('exp', None),
+            nbf=payload.get('nbf', None),
+        )
+        jwt.__decoded_payload = payload
+        jwt.__decoded_headers = headers
+        return jwt
+
+    @property
+    def payload(self):
+        if self.__decoded_payload:
+            return self.__decoded_payload
+        else:
```

Comment:
nit: I think it's usually clearer to write these functions that have two branches where the first is kind of a base case, short-circuity thing like

```python
@property
def payload(self):
    if self.__decoded_payload:
        return self.__decoded_payload

    payload = self._generate_payload().copy()
    # ...
    return payload
```